### PR TITLE
Map device to CUDAEventCache

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141338

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/cpp/c10d/ProcessGroupNCCLTest.cpp [hunk 1]
```diff
@@ -761,6 +761,7 @@ class ProcessGroupNCCLTest : public ::testing::Test {
   int size_{1};
 };
 
+/*
```
Comment (fduwjj) [inline test/cpp/c10d/ProcessGroupNCCLTest.cpp:764]: why are we doing this? The test is indeed needed.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -828,10 +829,29 @@ std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
...
+  auto it = cacheMap.find(index);
+  // If not, add a new entry
+  if (it == cacheMap.end()) {
+    // emplace uses in-place contruction, which avoids move or copy of the cache
+    // (the mutex of the cache is not movable)
+    it = cacheMap.emplace_hint(
+        it,
+        std::piecewise_construct,
+        std::forward_as_tuple(index),
+        std::forward_as_tuple()); // Default constructor of cache
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:851]: what is this.... this is cpp dark magic!

File: test/cpp/c10d/ProcessGroupNCCLTest.cpp (hunk 1 above)
Comment (kwen2501) [reply test/cpp/c10d/ProcessGroupNCCLTest.cpp:764]: I agree. This is just a draft PR trying the in-place constructor. I can copy the updated test in your PR if you'd like this PR to be a formal one. Let me know.

**Comment by kwen2501** [comment]
@fduwjj landed it in another PR.

**Pull request closed without merging**